File: src/markov/WeightedAdjacencyListGraph.java
package markov;

import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

public class WeightedAdjacencyListGraph<E> extends AdjacencyListGraph<E> implements WeightedGraph<E> {
    private Map<E, Map<E, Integer>> weights;

    public WeightedAdjacencyListGraph() {
        weights = new HashMap<>();
    }

    @Override
    public boolean add(E node) {
        return (super.add(node) && (weights.putIfAbsent(node, new HashMap<>()) == null));
    }

    @Override
    public void setWeight(E from, E to, int weight) {
        if (super.hasEdge(from, to)) weights.get(from).put(to, weight);
    }

    @Override
    public int getWeight(E from, E to) {
        return (super.hasEdge(from, to) ? weights.get(from).getOrDefault(to, 0) : 0);
    }

    @Override
    public Map<E, Integer> getNeighborWeights(E node) {
        var weights = new HashMap<E, Integer>();
        var neighborWeights = this.weights.get(node);
        if (null != neighborWeights) weights.putAll(neighborWeights);
        return weights;
    }


    @Override
    public String toString() {
        var ret = new ArrayList<String>();

        for (E node : getNodes()) {
            var neighborStrings = new ArrayList<String>();
            for (E neighbor : getNeighbors(node))
                neighborStrings.add(neighbor + "(" + getWeight(node, neighbor) + ")");

            ret.add(node + "->" + neighborStrings);
        }

        return ret.toString();
    }
}

File: src/markov_kt/WeightedAdjacencyListGraph.kt
package markov_kt

import markov.AdjacencyListGraph
import markov.WeightedGraph

class WeightedAdjacencyListGraph<E> : AdjacencyListGraph<E>(), WeightedGraph<E> {
    private val weights: MutableMap<E, MutableMap<E, Int>?>

    init {
        weights = HashMap()
    }

    override fun add(node: E): Boolean {
        return super.add(node) && weights.putIfAbsent(node, HashMap()) == null
    }

    override fun setWeight(from: E, to: E, weight: Int) {
        if (super.hasEdge(from, to)) weights[from]!![to] = weight
    }

    override fun getWeight(from: E, to: E): Int {
        return if (super.hasEdge(from, to)) weights[from]!!.getOrDefault(to, 0) else 0
    }

    override fun getNeighborWeights(node: E): Map<E, Int> {
        val weights = HashMap<E, Int>()
        val neighborWeights: Map<E, Int>? = this.weights[node]
        if (null != neighborWeights) weights.putAll(neighborWeights)
        return weights
    }

    override fun toString(): String {
        val ret = ArrayList<String>()
        for (node in nodes) {
            val neighborStrings = ArrayList<String>()
            for (neighbor in getNeighbors(node)) neighborStrings.add(
                neighbor.toString() + "(" + getWeight(
                    node,
                    neighbor
                ) + ")"
            )
            ret.add(node.toString() + "->" + neighborStrings)
        }
        return ret.toString()
    }
}